.github/workflows/deploy.yaml

```diff
@@ -65,3 +65,36 @@ jobs:
           tags: ${{ steps.meta.outputs.tags }}
           labels: ${{ steps.meta.outputs.labels }}
           context: .
+
+  monitor:
+    name: Build Monitor Lambda Zip
+    runs-on: ubuntu-22.04
+    permissions:
+      contents: write
+    env:
+      GH_TOKEN: ${{ github.token }}
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - uses: actions/setup-go@v5
+        with:
+          go-version: '>=1.22'
+      
+      - name: Build monitor
+        run: |
+          cd tools/autograph-monitor
+          make build
+      
+      - name: Create release draft if one doesn't exist yet
+        run: |
+          EXISTING_RELEASE=$(gh release view $GITHUB_REF_NAME --json="id" || echo '')
+          if [[ -z $EXISTING_RELEASE ]]; then
+              gh release create $GITHUB_REF_NAME --generate-notes -d
```

Comment:
If someone does make the git tag by going through GitHub's create release flow, wouldn't this never add the artifact to the release (or perhaps, depending on GH's implementation, race with the release)? 

And, if we fix that, isn't there a possibility of this creating a loop? This would create a draft release of our entire repo which would create a loop if run from someone creating a release with a new tag since these tasks are being run when they're `git tag`'ed, triggering this? (There's currently just extra workflow runs happening, I believe)

I've been trying to think through possible solutions and one of the additional problems is that we're also going to do a monitor and autograph-app staging deploy at the same time with this code and that each have different desires for how tags and GH releases work. What if we had separate tags for each of them?

I'm not sure that the release flow is the right solution for storage after we solve that problem, though. If we used the upload-artifact action, could we get consistent URLs based off our workflow runs that jenkins can use when told what to look for?

Reply:
> If someone does make the git tag by going through GitHub's create release flow, wouldn't this never add the artifact to the release (or perhaps, depending on GH's implementation, race with the release)?

If somebody makes the git tag by going through GitHub's release flow, this will just append the zip file to that existing release. If an existing release is not found (somebody created a tag without already creating a release), then it will create a draft release so it has something it can attach the zip file to. We can edit those release notes or delete them later before publishing.

I've updated the name of that step to be a little more clear.

> And, if we fix that, isn't there a possibility of this creating a loop?

This should not be able to create a loop in that way. At least I haven't been able to create that behavior in other projects with a similar workflow (ex: [kinto-admin](https://github.com/Kinto/kinto-admin/blob/master/.github/workflows/release.yml)).

> What if we had separate tags for each of them?

We could do that. But with shared code, creating release notes may not be intuitive to know which patches apply to which service. That may take some more thinking. Should we try to diagram this out?

> If we used the upload-artifact action, could we get consistent URLs based off our workflow runs that jenkins can use when told what to look for?

Workflow artifacts are a bit trickier than release files. There can only be one release with that name to pull data from. But for a given tag, you could have N workflows with Y artifacts to pull. We could pass that info over to jenkins in a webhook from the github action, but that doesn't work if we're running the job manually in Jenkins. We'd need to create some logic to identify which job we cared about more than one matches our criteria, and allow it to be overridden by passing in a job id.

Is this something we want to explore?

Reply:
I def misread these two steps.

I'm happy to ship this as-is and we can deal with the outcomes if they suck when they suck.

Reply:
I def misread these two steps.

I'm happy to ship this as-is and we can deal with the outcomes if they suck when they suck.